advanced tutorial (#221)
Oneflow-Inc/libai: docs/source/tutorials/Advanced_Tutorial.md

> # Advanced Tutorial

Reviewer: 感觉可以单独开一个advanced_tutorials文件夹，然后把这个教程放进去，可能后续还要加一些别的advanced教程

## Define your own parallel model with LiBai.layers

### Large-scale FC

> Let's say that you have a huge fully-connected-layer for large-scale classification (e.g., 1000w classes), which makes it impossible to fit into a single GPU.

Reviewer 2: ```suggestion
Suppose you have a huge fully-connected-layer for large-scale classification (e.g., 1000w classes), which makes it impossible to fit into a single GPU.
```

> Don't worry, with help of `LiBai.layers`, you can write models in a familiar way that you used to write models for a single GPU. We give a simple example showing how to write a tensor-parallel fully-connected-layer with 2 GPUs.

Reviewer: `with help of` -> `with the help of`，感觉这样比较合适

```python
# huge_fc_example.py
import oneflow as flow
from omegaconf import DictConfig
from oneflow import nn

from libai.layers import Linear
from libai.utils import distributed as dist

cfg = DictConfig(dict(data_parallel_size=1, tensor_parallel_size=2, pipeline_parallel_size=1))
dist.setup_dist_util(cfg)

class Huge_FC(nn.Module):
    def __init__(self):
        super().__init__()
        self.fc = Linear(2048, 32768, parallel="col")

    def forward(self, x):
        return self.fc(x)

huge_fc = Huge_FC()

x = flow.rand(32, 2048, sbp=flow.sbp.broadcast, placement=flow.placement("cuda", ranks=[0, 1]))
y = huge_fc(x)

print(f"rank: {flow.env.get_rank()}, tensor shape: {y.to_local().shape}")
```

You can run this toy example with command line as follows:

```shell
python3 -m oneflow.distributed.launch --nproc_per_node 2 huge_fc_example.py

>> rank: 0, tensor shape: oneflow.Size([32, 16384])
>> rank: 1, tensor shape: oneflow.Size([32, 16384])
```

Through the result, you can find that `y` has been split along with `axis=1` on 2 GPUs.

### Large MLP models

> MLP is very popular in transformer-based models. Assume we have a huge MLP model and its very large hidden size makes it difficult to fit into a single GPU.

Reviewer 3: ```suggestion
MLP is very popular in transformer-based models. Assume we have a huge MLP model with huge hidden size which makes it difficult to fit into a single GPU.
```

> We can then split the model weights across GPUs in a 2D mesh while you still write your model in a familiar way.

Reviewer 3: 下面的历史 模型的 sbp 是 [broadcast, split] ，感觉这里描述的权值 切割成 2d 网格形式有点不符？

cfg = DictConfig(dict(data_parallel_size=2, tensor_parallel_size=2, pipeline_parallel_size=1))
dist.setup_dist_util(cfg)

# Write a Simple 2D Parallel MLP
class MLP_2D(nn.Module):
    def __init__(self):
        super().__init__()
        self.linear_1 = Linear(in_features=1024, out_features=16384, parallel="col")
        self.relu = nn.GELU()
        self.linear_2 = Linear(in_features=16384, out_features=1024, parallel="row")

    def forward(self, x):
        x = self.linear_1(x)
        x = self.relu(x)
        x = self.linear_2(x)
        return x

# define a model
mlp = MLP_2D()

# define input with 2D sbp
x = flow.rand(
    32,
    1024,
    sbp=dist.get_nd_sbp([flow.sbp.split(0), flow.sbp.broadcast]),
    placement=dist.get_layer_placement(0)
)
y = mlp(x)

print(f"rank: {flow.env.get_rank()}, tensor shape: {y.to_local().shape}")
```

You can run it with command line as follows:

```shell
python3 -m oneflow.distributed.launch --nproc_per_node 4 huge_mlp_example.py

>> rank: 2, tensor shape: oneflow.Size([16, 1024])
>> rank: 3, tensor shape: oneflow.Size([16, 1024])
>> rank: 1, tensor shape: oneflow.Size([16, 1024])
>> rank: 0, tensor shape: oneflow.Size([16, 1024])
```

> From above, you can see that data are split into 2 groups for data parallel and weights are split into 2 groups for model parallel. So this simple example just implements a 2D parallel.

Reviewer 2: 上文用的 `tensor parallel` 这里用的`model  parallel`, 最好统一一下

Reviewer: 是不是要有什么地方再强调一下模型并行是tensor parallel, 我发现官网叫model parallel, 但是我们内部叫tensor parallel 😆

Author: 我写成 tensor model parallel 吧

Reviewer 4: 这里的2D并行看起来不像是colossal ai的2d tensor parallel，求问为什么叫2D

Author: 数据并行+模型并行，所以叫 2d 并行

## Write your own pipeline parallel model

In this tutorial, you will learn how to use pipeline parallel in your own model. In LiBai, we have two pipeline-parallel modes: naive pipeline parallel and (similar) 1F1B pipeline parallel introduced by [Megatron-LM](https://arxiv.org/abs/1909.08053).

### Introduction of naive pipeline parallel

In LiBai, naive pipeline parallel can be implemented by setting layers and parameters `placement`. 
You can easily configure their `placement` by `dist.get_layer_placement(idx)`.

We give an example for `placement` configuration.

```python
# set a free tensor placement to first stage
self.pos_embed = nn.Parameter(
    flow.zeros(
        1,
        num_patches + 1,
        embed_dim,
        sbp=dist.get_nd_sbp([flow.sbp.broadcast, flow.sbp.broadcast]),
        placement=dist.get_layer_placement(0),
    )
)

# set a Linear placement to last stage
# set it manually 
self.head = Linear(embed_dim, num_classes, layer_idx=-1).to_global(placement=dist.get_layer_placement(-1))
# use `layer_idx` API
self.head = Linear(embed_dim, num_classes, layer_idx=-1)
```

> After configuring models placement, you need to add the input placement transition across different stages. In LiBai, we set a `layer_idx` attribute in each `nn.Module`, so you can simply add `to_global` in `forward` to implement input placement transition.

Reviewer 3: ```suggestion
After configuring model‘s placement, you need to add the input placement transition across different stages. In LiBai, we set a `layer_idx` attribute in each `nn.Module`, so you can simply add `to_global` in `forward` to implement input placement transition.
```
